File: ios/HybridNitroContacts.swift
import Foundation
import NitroModules
import Contacts

class HybridNitroContacts: HybridNitroContactsSpec {
    let contactService: ContactServiceDelegate = ContactService()
    
    func getContacts() throws -> Promise<[NitroContact]>{
        return .async{
            do {
                return try await self.contactService.getAll()
            } catch {
                throw error
            }
        }
    }
    
    func getContact(id: String) throws -> Promise<NitroContact?>{
        return .async{
            do {
                return try await self.contactService.getById(contactId: id)
            } catch {
                throw error
            }
        }
    }
    
    func requestPermission() throws -> Promise<Bool> {
        return .async {
            do {
                return try await CNContactStore().requestAccess(for: .contacts)
            } catch {
                throw error
            }
        }
    }
    
    func authorizationStatus() throws -> Promise<NitroAuthorizationStatus> {
        return .async {
            let status = CNContactStore.authorizationStatus(for: .contacts)
            switch status {
            case .notDetermined:
                return .notdetermined
            case .authorized:
                return .authorized
            case .denied:
                return .denied
            case .restricted:
                return .restricted
            case .limited:
                return .limited
            @unknown default:
                fatalError()
            }
        }
    }
}

File: android/src/main/java/com/nitrocontacts/HybridNitroContacts.kt
package com.nitrocontacts

import android.util.Log
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.margelo.nitro.nitrocontacts.HybridNitroContactsSpec
import com.margelo.nitro.nitrocontacts.NitroAuthorizationStatus
import com.margelo.nitro.nitrocontacts.NitroContact

class HybridNitroContacts: HybridNitroContactsSpec() {
    private val context = NitroModules.applicationContext
    private val contactService = NitroContactService(context)

    override fun getContacts(): Promise<Array<NitroContact>> {
       return Promise.async {
           try {
               contactService.getAll()
           } catch (error: Exception) {
               Log.d(TAG, error.message.toString())
               throw Error(error)
           }
       }
    }

    override fun getContact(id: String): Promise<NitroContact?> {
        return Promise.async {
            try {
                contactService.getById(id)
            } catch (error: Exception) {
                Log.d(TAG, error.message.toString())
                throw Error(error)
            }
        }
    }

    override fun authorizationStatus(): Promise<NitroAuthorizationStatus> {

        // IOS only always returns AUTHORIZED
        return Promise.async {
            NitroAuthorizationStatus.AUTHORIZED
        }
    }

    override fun requestPermission(): Promise<Boolean> {
        return Promise.async {
            true
        }
    }

    companion object {
        const val TAG = "HybridNitroContacts"
    }
}
